### PR TITLE
Models for Unloading Permission

### PR DESCRIPTION
**Pull request by TheNickWilson**

### CODE REVIEW

File: test/models/PackageSpec.scala [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+          val expectedPackage = UnpackedPackage(kindOfPackage, numberOfPieces, None)
+
+          json.validate[Package] mustEqual JsSuccess(expectedPackage)
+      }
+    }
+
+    "must deserialise to a Regular Package" in {
+
+      val gen = for {
+        kindOfPackage    <- stringWithMaxLength(3)
```
Comment (amaalali) [inline test/models/PackageSpec.scala]: Would changing this to `Gen.oneOf(BulkPackage.validCodes)` mean we can remove the `whenever` below?

File: test/models/RegularPackageSpec.scala [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+import play.api.libs.json.{JsError, JsSuccess, Json}
+
+class RegularPackageSpec extends FreeSpec with MustMatchers with ScalaCheckPropertyChecks with ModelGenerators {
+
+  "Regular Package" - {
+
+    "must deserialise when the kind of package does not indicate Bulk or Unpacked" in {
+
+      val gen = for {
+        kindOfPackage    <- stringWithMaxLength(3)
```
Comment (amaalali) [inline test/models/RegularPackageSpec.scala]: Similar question here as above

File: test/models/TraderAtDestinationSpec.scala [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+          (eori, name, streetAndNumber, postCode, city, countryCode)  =>
+
+            val json = Json.obj(
+              "eori"            -> eori,
+              "name"            -> name,
+              "streetAndNumber" -> streetAndNumber,
+              "postCode"        -> postCode,
+              "city"            -> city,
+              "countryCode"     -> countryCode
+            )
```
Comment (amaalali) [inline test/models/TraderAtDestinationSpec.scala]: Would it be better to generate an `arbitrary[TraderAtDestinationWithEori]` and then use the fields from that to construct the json?

File: app/models/Enumerable.scala [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+
+    implicit def pathBindable[A](implicit ev: Enumerable[A]): PathBindable[A] =
+      new PathBindable[A] {
+
+        override def bind(key: String, value: String): Either[String, A] =
+          ev.withName(value).toRight("error.invalid")
+
+        override def unbind(key: String, value: A): String =
+          value.toString
+      }
```
[the remark this answers is not in the document]
Comment (TheNickWilson) [reply app/models/Enumerable.scala]: Good spot.  I don't have a strong preference but expect implicit parameters will be more familiar to most, so I'll change the `writes` to match

File: app/models/SpecialMention.scala [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    implicit def convertToSupertype[A, B >: A](a: Reads[A]): Reads[B] =
+      a.map(identity)
+
+    SpecialMentionEc.reads or
+      SpecialMentionNonEc.reads
+  }
+
+  implicit lazy val writes: Writes[SpecialMention] = Writes {
```
[the remark this answers is not in the document]
Comment (TheNickWilson) [reply app/models/SpecialMention.scala]: Yup - I've gone through the other models as well and standardised on OWrites. Thanks!

File: app/models/WithName.scala [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package models
+
+class WithName(string: String) {
```
[the remark this answers is not in the document]
Comment (TheNickWilson) [reply app/models/WithName.scala]: :+1:

File: test/models/PackageSpec.scala (hunk 1 above)
Comment (TheNickWilson) [reply test/models/PackageSpec.scala]: This test is for a `RegularPackage`, so it needs to have any `kindOfPackage` that _isn't_ a valid bulk or unpacked code.  Short of specifying all valid "regular" codes I don't think we can avoid the whenever

File: test/models/RegularPackageSpec.scala (hunk 2 above)
Comment (TheNickWilson) [reply test/models/RegularPackageSpec.scala]: Same as above, this is for a `RegularPackage`

File: test/models/TraderAtDestinationSpec.scala (hunk 3 above)
Comment (TheNickWilson) [reply test/models/TraderAtDestinationSpec.scala]: The issue is that TraderAtDestinationWithEori has lots of optional fields, so building the json from a model means mapping over each optional field and gets harder to read IMO

**Pull request merged**